docs/proposals/202105256-cluster-class-and-managed-topologies.md

```diff
@@ -229,8 +262,182 @@ type LocalObjectTemplate struct {
 }
 ```
 
+**VariablesClass**
+
+```golang
+// VariablesClass defines the variables which can be configured
+// in the Cluster topology and used in patches.
+type VariablesClass struct {
+  // Definitions define the variables inline.
+  Definitions []VariableDefinitionClass `json:"definitions,omitempty"`
+}
+ 
+// VariableDefinitionClass defines a variable which can 
+// be configured in the Cluster topology and used in patches.
+type VariableDefinitionClass struct {
+  // Name of the variable.
+  Name string `json:"name"`
+ 
+  // Required specifies if the variable is required.
+  // Note: this applies to the variable as a whole and thus the 
+  // top-level object defined in the schema. If nested fields are 
+  // required, this will be specified inside the schema.
+  Required bool `json:"required"`
+ 
+  // Schema defines the schema of the variable.
+  Schema VariableDefinitionSchemaClass `json:"schema"`
+}
+ 
+// VariableDefinitionSchemaClass defines the schema of a variable.
+type VariableDefinitionSchemaClass struct{
+  // OpenAPIV3Schema defines the schema of a variable via OpenAPI v3
+  // schema. The schema is a subset of the schema used in
+  // Kubernetes CRDs.
+  OpenAPIV3Schema JSONSchemaProps `json:"openAPIV3Schema"`
+}
+```
+
+The schema implementation will be built on top of the [Open API schema embedded in Kubernetes CRDs](https://github.com/kubernetes/apiextensions-apiserver/blob/master/pkg/apis/apiextensions/types_jsonschema.go).
+To keep the implementation as easy and user-friendly as possible we will only implement the following feature set 
+for now (until further use cases emerge):
+- Basic types:
+    - boolean, integer, number, string
+    - nullable, i.e. each of the basic types can be set to null
+- Basic validation, e.g. format, minimum, maximum, pattern, required, ...
+- Defaulting
+    - Defaulting will be implemented based on the CRD structural schema library and thus will have the same feature set 
```

Comment:
can we consider allowing built in variables and other variable names as defaults in the future? For example, an Azure vnet name defaults to `{{ .builtin.cluster.name }}-vnet` if not specified.

Reply:
It's not supported in [JSON Schema](https://json-schema.org/understanding-json-schema/reference/generic.html?highlight=default) nor in [OpenAPI Schema v3](https://swagger.io/docs/specification/data-models/keywords/?sbsearch=default) and thus also not in CRDs.

Additional context:
* the `Default` field in CRDs is a [apiextensionsv1.JSON](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1/types_jsonschema.go#L59) (essentially a byte array)
* the current plan is to re-use the CRD [structural schema defaulting package](https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/defaulting) to avoid reimplementing

Independent of that, I think we could implement it, if we want to. A few variants:

#### 1. Constant default values without variables
* could be implemented by just using the CRD defaulting package

#### 2. default values with builtin variables
* we could add a pre-processing step which does go template evaluation over the JSON value
* the result is then used as default value for the CRD defaulting package
Notes:
* Not sure if in this case we should just make the existing `Default` field a template or introduce something like `DefaultTemplate` (which doesn't exist at all in JSON Schema / Open API Schema v3)
* Validation on the ClusterClass wouldn't be entirely trivial, but we could consider something similar to how we would validate variable templates in patches.
* I think we should restrict this to "global" builtin variables which are not "context-sensitive", i.e. `cluster.{name,namespace,labels,annotations}`, `cluster.topology.{version,class}`. As we (currently) only have "global" variables in the Cluster, I think that would make sense.
    * Maybe we should restrict this even further to a subset of those, e.g. labels and annotations don't really make sense as long as we only have basic types, and I'm not sure if we really need them.

#### 3. default values with built-in and user-defined variables

Similar to 2. with the following additional considerations:
* we have to avoid circular dependencies between default values of variables
   * or evaluate them like envsubst: one after another in order of the array and when a referenced variable is not set yet, use the default value
* Validation would be similar as above, but even more brittle

Reply:
Agree, it would be very useful. We're currently using webhooks for such kind of defaulting, but having it built-in would definitely be preferred.

Reply:
FYI, we are just starting to explore external patches, that will be the next evolution of this proposal.
In one of those discussions we considered the idea of allowing external patches to provide additional variables, including their schema and default value.

This has some complexities still to be addressed, but also some clear benefits that can be relevant for this thread:
- External patches will be free to add dynamic defaulting rules, addressing the use case described above (more flexibility)
- From an UX point of view, everything will be simpler. Just add an external patch to a ClusterClass, and then the ClusterClass will get automatically new variables, without the need of manually writing Open API schema definition for them (less yaml)
- Variables will be implicitly aligned to the external patch version (less maintenance)

Generalising a little bit, I personally consider inline patches a no-code, ~simple solution for a limited number of use cases; we can extend it, but we will always hit some limits no matter how much we stretch OpenAPI Schema or JSON Patches standards. Instead, external patches, at least in theory, are a better solution for advanced users, given that they will provide full control on variables/generated patches and the power of real programming languages. 

I'm more than happy to discuss about it if you are interested in this topic